### ISSUE TITLE
About cluster_addons variable specs

### ISSUE THREAD

**Issue by igoratencompass** [terraform-aws-modules/terraform-aws-eks#2427]
## Description

Hi all, first would like to point out that this is not a feature request but more a documentation improvement request.

Maybe a silly question but when going to the [cluster_addons](https://github.com/terraform-aws-modules/terraform-aws-eks#input_cluster_addons) parameter it is not clear, to me at least, where to go from here. Where do we find the info about which add-ons are supported and their config parameters we can pass in (if any)? Are those parameters all the same or different? Are they related to the options supported as per the AWS driver documentation? Or they are just based on `aws_eks_addon` resource so we should look there?

Apart from hunting them down across examples like this one: 
```
  cluster_addons = {
    coredns = {
      preserve    = true
      most_recent = true

      timeouts = {
        create = "25m"
        delete = "10m"
      }
    }
    kube-proxy = {
      most_recent = true
    }
    vpc-cni = {
      most_recent = true
    }
  }
```
I don't see any other option. And it is further not clear if the example shows all or just part of the options about the given add-on. Is this all that the `vpc-cni` itself for example supports in terms of parameters? And if not what are the other parameters supported? Just to find in another example:

```
    vpc-cni = {
      most_recent              = true
      service_account_role_arn = module.vpc_cni_irsa.iam_role_arn
      configuration_values = jsonencode({
        env = {
          # Reference docs https://docs.aws.amazon.com/eks/latest/userguide/cni-increase-ip-addresses.html
          ENABLE_PREFIX_DELEGATION = "true"
          WARM_PREFIX_TARGET       = "1"
        }
      })
    }
```

Then I can find another piece of info in another example that I can use:

```
resolve_conflicts = "OVERWRITE"
```
ok, is this one exclusive with `preserve = true` or there is another meaning to it?

I think it would benefit everyone if this input variable is further broken down into its building blocks with full specs per add-on instead scattered around examples. But on the other hand maybe it is just me in which case you can ignore this.

Thanks for all your great work plus the time invested in reading this.

BTW, I was also looking at https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/docs/irsa_integration.md and wonder isn't this introducing circular dependency in terraform? As per the example:

```
module "eks" {
...
  cluster_addons = {
    vpc-cni = {
      resolve_conflicts        = "OVERWRITE"
      service_account_role_arn = module.vpc_cni_irsa.iam_role_arn
    }
  }
...
}

module "vpc_cni_irsa" {
...
  oidc_providers = {
    main = {
      provider_arn               = module.eks.oidc_provider_arn
      namespace_service_accounts = ["kube-system:aws-node"]
    }
  }
...
}
``` 
the `service_account_role_arn` in the `module.eks` depends on `module.vpc_cni_irsa` and the `module.vpc_cni_irsa` depends on `module.eks` ... is this really working?

**Comment by bryantbiggs** [comment]
See https://aws.amazon.com/blogs/containers/amazon-eks-add-ons-advanced-configuration/

You can query the API to get the schema of the configuration:
```sh
aws eks describe-addon-configuration --addon-name vpc-cni --addon-version v1.12.0-eksbuild.1 --query 'configurationSchema' --output text | jq
```

> the service_account_role_arn in the module.eks depends on module.vpc_cni_irsa and the module.vpc_cni_irsa depends on module.eks ... is this really working?

No - Terraform is able to resolve the individual pieces of information it needs in the correct order without issue

**Comment by prasoon-pxc** [comment]
I also wanted to add this addons on my existing cluster and was looking for doc, which tells , what will be the impact of it on existing cluster

**Comment by obeleh** [comment]
The json schema was: 
```
{
  "$ref": "#/definitions/VpcCni",
  "$schema": "http://json-schema.org/draft-06/schema#",
  "definitions": {
    "Env": {
      "additionalProperties": false,
      "properties": {
        "ADDITIONAL_ENI_TAGS": {
          "type": "string"
        },
        "ANNOTATE_POD_IP": {
          "format": "boolean",
          "type": "string"
        },
        "AWS_VPC_CNI_NODE_PORT_SUPPORT": {
          "format": "boolean",
          "type": "string"
        },
        "AWS_VPC_ENI_MTU": {
          "format": "integer",
          "type": "string"
        },
        "AWS_VPC_K8S_CNI_CUSTOM_NETWORK_CFG": {
          "format": "boolean",
          "type": "string"
        },
        "AWS_VPC_K8S_CNI_EXCLUDE_SNAT_CIDRS": {
          "type": "string"
        },
        "AWS_VPC_K8S_CNI_EXTERNALSNAT": {
          "format": "boolean",
          "type": "string"
        },
        "AWS_VPC_K8S_CNI_LOGLEVEL": {
          "type": "string"
        },
        "AWS_VPC_K8S_CNI_LOG_FILE": {
          "type": "string"
        },
        "AWS_VPC_K8S_CNI_RANDOMIZESNAT": {
          "type": "string"
        },
        "AWS_VPC_K8S_CNI_VETHPREFIX": {
          "type": "string"
        },
        "AWS_VPC_K8S_PLUGIN_LOG_FILE": {
          "type": "string"
        },
        "AWS_VPC_K8S_PLUGIN_LOG_LEVEL": {
          "type": "string"
        },
        "CLUSTER_ENDPOINT": {
          "type": "string"
        },
        "DISABLE_INTROSPECTION": {
          "format": "boolean",
          "type": "string"
        },
        "DISABLE_METRICS": {
          "format": "boolean",
          "type": "string"
        },
        "DISABLE_NETWORK_RESOURCE_PROVISIONING": {
          "format": "boolean",
          "type": "string"
        },
        "ENABLE_BANDWIDTH_PLUGIN": {
          "format": "boolean",
          "type": "string"
        },
        "ENABLE_NFTABLES": {
          "format": "boolean",
          "type": "string"
        },
        "ENABLE_POD_ENI": {
          "format": "boolean",
          "type": "string"
        },
        "ENABLE_PREFIX_DELEGATION": {
          "format": "boolean",
          "type": "string"
        },
        "ENI_CONFIG_ANNOTATION_DEF": {
          "type": "string"
        },
        "ENI_CONFIG_LABEL_DEF": {
          "type": "string"
        },
        "INTROSPECTION_BIND_ADDRESS": {
          "type": "string"
        },
        "MAX_ENI": {
          "format": "integer",
          "type": "string"
        },
        "MINIMUM_IP_TARGET": {
          "format": "integer",
          "type": "string"
        },
        "POD_SECURITY_GROUP_ENFORCING_MODE": {
          "type": "string"
        },
        "WARM_ENI_TARGET": {
          "format": "integer",
          "type": "string"
        },
        "WARM_IP_TARGET": {
          "format": "integer",
          "type": "string"
        },
        "WARM_PREFIX_TARGET": {
          "format": "integer",
          "type": "string"
        }
      },
      "title": "Env",
      "type": "object"
    },
    "Init": {
      "additionalProperties": false,
      "properties": {
        "env": {
          "$ref": "#/definitions/InitEnv"
        }
      },
      "title": "Init",
      "type": "object"
    },
    "InitEnv": {
      "additionalProperties": false,
      "properties": {
        "DISABLE_TCP_EARLY_DEMUX": {
          "format": "boolean",
          "type": "string"
        }
      },
      "title": "InitEnv",
      "type": "object"
    },
    "Limits": {
      "additionalProperties": false,
      "properties": {
        "cpu": {
          "type": "string"
        },
        "memory": {
          "type": "string"
        }
      },
      "title": "Limits",
      "type": "object"
    },
    "Resources": {
      "additionalProperties": false,
      "properties": {
        "limits": {
          "$ref": "#/definitions/Limits"
        },
        "requests": {
          "$ref": "#/definitions/Limits"
        }
      },
      "title": "Resources",
      "type": "object"
    },
    "VpcCni": {
      "additionalProperties": false,
      "properties": {
        "env": {
          "$ref": "#/definitions/Env"
        },
        "init": {
          "$ref": "#/definitions/Init"
        },
        "resources": {
          "$ref": "#/definitions/Resources"
        }
      },
      "title": "VpcCni",
      "type": "object"
    }
  }
}
```

And this was the configuration value we used:

```
{
	"env": {
    	"ENABLE_PREFIX_DELEGATION": "true",
        "WARM_PREFIX_TARGET": "1"
    }	
}
```

**Comment by vuskeedoo** [comment]
I am also trying to apply custom configuration values to the VPC CNI. I tried using jsonencode and string value, but the env vars do not apply to the aws-node daemon set.

```
cluster_addons = {
    vpc-cni = {
      addon_version = "v1.12.1-eksbuild.1"
      resolve_conflicts = "OVERWRITE"
      service_account_role_arn = module.vpc_cni_ipv4_irsa_role.iam_role_arn
      configuration_values = jsonencode({
        env = {
          WARM_IP_TARGET = "11"
          MINIMUM_IP_TARGET = "20"
        }
      })
    }
}

Example 2: "{\"Env\":{\"name\":\"WARM_IP_TARGET\",\"value\":\"11\"},{\"name\":\"MINIMUM_IP_TARGET\",\"value\":\"20\"}}"

**Comment by gw0** [comment]
@vuskeedoo Have you tried this recommended workaround for `vpc-cni` with ENABLE_PREFIX_DELEGATION:

- https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/examples/ipv4-prefix-delegation

**Comment by fstr** [comment]
@vuskeedoo please make sure that you have at least version `19.1.0` of the `terraform-aws-eks` module.

This functionality was added in 19.1.0. See [changelog](https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/CHANGELOG.md#1910-2022-12-16).

**Comment by Raboo** [comment]
Hi, I googled my way to this issue. I feel that some documentation on how to list available addons and their respective parameters would help a lot.
For instance I am using terraform [AWS EKS Blueprints](https://aws-ia.github.io/terraform-aws-eks-blueprints/main/) to setup aws-ebs-csi-driver, I just noticed that it's possible to setup via the `cluster_addons = {}` block. Also I don't really know if there are any pros or cons doing one way or another.
